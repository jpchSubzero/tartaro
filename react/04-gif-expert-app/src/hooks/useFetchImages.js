import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

//Snippet => rafc y luego borrar el import porque no se necesita react para JSX y quitar el return
// export const useFetchImages = () => {

//Recibir la categoría
export const useFetchImages = (category) => {
        //Se usa state porque el hook no sabe para que va a trabajar y por convención se usa state en lugar de otro nombre de variable
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    //Agregar el setTimeout para mostrar como cambia el mensaje cargado
    // setTimeout( () =>  {
    //     setState({
    //         data: [1,2,3,4,5,6],
    //         loading: false    
    //     });
    // }, 3000);

    // Comentar el setTimeout
    //Agergar el useEffect para controlar que solo se ejecute cuando queremos, en esta ocasión cuando cambia la categoría
    //Los useEffect no pueden ser async, siempre esperan un valor síncrono
    useEffect(() => {

            getGifs(category).then(resolve => {

                //Agregarmos el setTimeout solo para ver como se ejecuta
                //Comentar al final y solo hacer el useState
                // setTimeout( () =>  {
                //     console.log(resolve);
                //     setState({
                //         data: resolve,
                //         loading: false
                //     });
                // }, 3000);

                setState({
                    data: resolve,
                    loading: false
                });

            });
    }, [category]);


    return state;
}
