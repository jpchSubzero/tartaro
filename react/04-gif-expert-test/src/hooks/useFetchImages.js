import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

// Se puede validar que al inicio exista data vacÃ­a y loading en true. TambiÃ©n que luego del effect exista data y loading en false.

export const useFetchImages = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
            getGifs(category).then(resolve => {
                setState({
                    data: resolve,
                    loading: false
                });

            });
    }, [category]);

    return state;
}
