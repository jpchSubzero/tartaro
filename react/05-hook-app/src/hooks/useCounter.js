// snippet -> rafc y quitar lo que no sirve
import { useState } from 'react';

// Creamos un hook que tiene un valor por defecto
export const useCounter = ( initialState = 10 ) => {
    const [state, setState] = useState(initialState);

    // const increment = () => {

    // Agregar un factor de suma
    const increment = (factor = 1) => {
        // setState(state + 1);

        // Cambiar el 1 por el factor
        setState(state + factor);
    };

    // const decrement = (factor = 1) => {

    // Agregar un factor de resta
    const decrement = (factor = 1) => {
        // setState(state - 1);

        // Cambiar el 1 por el factor
        setState(state - factor);
    };

    // Resoluci贸n de tarea. Agregar funci贸n de reset
    const reset = () => {
        setState(initialState);
    };
    
    // Hacemos accesibles las propiedades y funciones del hook
    return {
        state,
        increment,
        decrement,

        // Resoluci贸n de tarea. Devolver funci贸n de reset
        reset
    };
}
