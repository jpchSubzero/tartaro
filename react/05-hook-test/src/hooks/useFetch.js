import { useState, useEffect, useRef } from 'react';

export const useFetch = ( url ) => {
    if (!url) {
        throw new Error('No hay url');
    }

    const isMounted = useRef(true);

    useEffect(() => {
        return () => {
            isMounted.current = false;
        };
    }, []);

    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        setState({data: null, loading: true, error: null});

        fetch(url)
            .then(response => response.json())
            // Quitar el ; y agregar el catch para probar el error
            .then(data => {
                setTimeout(() => {
                    if (isMounted.current) {
                        setState({
                            loading: false,
                            error: null,
                            data
                        });                        
                    }
                }, 2000);
            })
            .catch(error => {
                setState({
                    loading: false,
                    error: 'Error al llamar al API: ' + error,
                    data: null                    
                });
            });
    }, [url]);

    return state;
}
