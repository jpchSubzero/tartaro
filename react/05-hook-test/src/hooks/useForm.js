// Este hook sirve para manejar los valores de elementos de formularios
import { useState } from 'react';

export const useForm = (initialState = {}) => {
    const [values, setValues] = useState(initialState);

    // Se puede devolver directamente el state y la función para manejar el evento pero eso hace que el return se vuelva un poco confuso por lo cual resulta más claro declarar la función
    // return [ 
    //     values, 
    //     ({ target }) => {
    //         setValues({ 
    //             ...values,
    //             [target.name]: target.value
    //         });        
    //     }
    // ];

    // Declara la función para que el return sea un poco más fácil de comprender
    const handleInputChange = ({ target }) => {
        setValues({ 
            ...values,
            [target.name]: target.value
        });        
    };

    // Agregar la funcionalidad de reset (limpiar campo) cuando se trabaje con los reducers en el ejercicio de TODO. En este caso no sirve el initialState.
    // Hay que hacerlo accesible retornándolo junto con values, handleInputChange
    const handleReset = () => {
        setValues(initialState);        
    };


    return [values, handleInputChange, handleReset];
}
