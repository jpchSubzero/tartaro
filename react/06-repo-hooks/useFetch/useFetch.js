import { useState, useEffect, useRef } from 'react';

export const useFetch = ( url ) => {
    if (!url) {
        throw new Error('No hay url');
    }

    // Agregar useRef para validar si el componente está disponible (montado) o no
    const isMounted = useRef(true);

    // Utilizamos el useEffect para determinar si el componente sigue disponible. Si fue desmontado llama a su función de retorno y ahí es donde cambiamos el ref. Si el componente sigue montado no se llama la función
    useEffect(() => {
        return () => {
            isMounted.current = false;
        };
    }, []);

    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        // Al final agregar para hacer que se presente el loading cada vez que se solicite una nueva frase
        setState({data: null, loading: true, error: null});

        fetch(url)
            .then(response => response.json())
            .then(data => {
                // setState({
                //     loading: false,
                //     error: null,
                //     // Como el campo y la variable de la promesa se llaman data no es necesario data: data ya que js asume que se autoasigna
                //     data
                // });

                // Agregar un timeOut para el ejercicio RealExampleRef y hacer que demore en responder
                // setTimeout(() => {
                //     setState({
                //         loading: false,
                //         error: null,
                //         data
                //     });
                // }, 2000);

                // Validar si el componente está montado para devolver la información
                setTimeout(() => {
                    if (isMounted.current) {
                        setState({
                            loading: false,
                            error: null,
                            data
                        });                        
                    }
                }, 2000);

            });
    }, [url]);

    return state;
}
