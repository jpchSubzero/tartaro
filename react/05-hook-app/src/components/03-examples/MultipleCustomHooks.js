import React from 'react';

// Importar el hook useFetch
import { useFetch } from '../../hooks/useFetch';

// Importar el hook useCounter
import { useCounter } from '../../hooks/useCounter';
import './styles.css';

export const MultipleCustomHooks = () => {
    // Invocamos el custom hook. El nombre state no es importante
    // https://breakingbadapi.com/documentation
    // const state = useFetch(`https://www.breakingbadapi.com/api/quotes/1`);
    // console.log(state);

    // Agregar el hook de contador para ir incrementando el id de la cita. Esto se hace después de agregar el botón
    const { state: counter, increment } = useCounter(1);

    // Destructuramos el state para leer sus propiedades
    // const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/1`);

    // Cambiar el id quemado por el counter
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

    // No se puede leer directamente la data porque en un inicio es null ya que es una promesa, por lo tanto hay que validarla
    // data puede ser null y para utilizarla como operador lógico hay que convertirlo en boolean, para eso se utiliza !, al negar null se convierte en true pero para evitar que se evalue la segunda condición, data[0], es necesario que sea false y para eso se agrega otro !. Traducido seria: si no hay data no obtienes datos de la posición 0, es decir, ni siquiera lo intenta; si existe data obtén la data de la posición 0
    // Esto se puede revisar escribiendo en la consola
    const { author, quote } = !!data && data[0];
    console.log(author, quote);

    return (
        <>
            {/* <h1>Custom Hooks</h1>    */}

            {/* Cambiar el nombre por el API que vamos a utilizar */}
            <h1>Breaking Bad Quotes</h1>
            <hr />

            {/* <div className="alert alert-info text-center">
                Cargando...
            </div> */}

            {/* https://getbootstrap.com/docs/5.0/content/typography/#alignment */}
            {/* <figure className="text-end">
                <blockquote className="blockquote">
                    <p>Frase</p>
                </blockquote>
                <figcaption className="blockquote-footer">
                    Nombre
                </figcaption>
            </figure> */}

            {/* Agregamos la validación para presentar el loading o la cita a través de un operador ternario */}
            {
                loading ? (
                    <div className="alert alert-info text-center">
                        Cargando frase...
                    </div>
                ) : (
                    <figure className="text-end">
                        <blockquote className="blockquote">
                            {/* <p>Frase</p> */}

                            {/* Asignar la cita */}
                            <p>{ quote }</p>
                        </blockquote>
                        <figcaption className="blockquote-footer">
                            {/* Nombre */}

                            {/* Asignar el autor */}
                            { author }
                        </figcaption>

                        {/* Se utiliza () => increment() porque recibe un parámetro opcional y aunque no se envíe es necesario indicar que puede hacerlo */}
                        <button className="btn btn-primary" onClick={() => increment()}>Siguiente frase</button>
                    </figure>
                )
            }
        </>
    )
}
