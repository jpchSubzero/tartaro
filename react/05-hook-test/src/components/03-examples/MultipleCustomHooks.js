import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import './styles.css';

export const MultipleCustomHooks = () => {
    const { state: counter, increment } = useCounter(1);
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    const { author, quote } = !!data && data[0];

    return (
        <>
            <h1>Breaking Bad Quotes</h1>
            <hr />
            <h3>{ counter }</h3>

            {
                loading ? (
                    <div className="alert alert-info text-center">
                        Cargando frase...
                    </div>
                ) : (
                    <figure className="text-end">
                        <blockquote className="blockquote">
                            <p className="mb-0">{ quote }</p>
                        </blockquote>
                        <figcaption className="blockquote-footer">
                            { author }
                        </figcaption>

                        <button className="btn btn-primary" onClick={() => increment()}>Siguiente frase</button>
                    </figure>
                )
            }
        </>
    )
}
