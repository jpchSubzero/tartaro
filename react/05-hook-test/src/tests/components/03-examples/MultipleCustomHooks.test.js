import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

// Permite realizar implementaciones falsas
jest.mock('../../../hooks/useFetch');

// Importar después de probar con useFetch
jest.mock('../../../hooks/useCounter');

describe('Pruebas con MultipleCustomHooks', () => {
    // Sin beforeEach presenta el error: TypeError: Cannot destructure property 'state' of '(0 , _useCounter.useCounter)(...)' as it is undefined.
    // En cada test creamos el snapshot, el beforeEach se ejecuta antes que se realice cada uno de ellos. Si utilizamos useCounter.mockReturnValue en cada prueba ya no muestra ese error
    beforeEach( () => {
        useCounter.mockReturnValue({
            state: 10,
            increment: () => {},

            // Podemos omitirlas porque no las estamos utilizando al inicio
            // decrement: () => {},
            // reset: () => {},
        });
    });

    test('Debe mostrarse correctamente', () => {
        // Luego de agregar jest.mock en la segunda prueba, agregar el mockeo del hook useFetch con la información que esperamos recibir, en este caso data, loading y error. En este caso no importa la data ya que para eso se probó el hook useFetch en otra prueba
        useFetch.mockReturnValue({
            loading: true,
            error: null,
            data: null              
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrar información', () => {
        // Al agregar el jest.mock se produce un error: 
        // TypeError: Cannot destructure property 'loading' of '(0 , _useFetch.useFetch)(...)' as it is undefined. 
        // En el componente en la parte de:
        // const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
        // Esto es porque al mockear el hook useFetch ya no utiliza la implementación del componente y debemos proveerle una nueva. jest.mock afecta a todo el archivo de pruebas y hay que corregir en la anterior
        useFetch.mockReturnValue({
            loading: false,
            error: null,
            data: [{
                quote_id: 1,
                quote: "Frase",
                author: "Author",
                series: "Serie"
            }]
        });
        
        const wrapper = shallow(<MultipleCustomHooks />);
        console.log(wrapper.html());

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert').exists()).toBe(false);

        // Agregar al p de quote la clase className="mb-0" para poder rastrear su contenido en la prueba
        expect(wrapper.find('.mb-0').text().trim()).toBe('Frase');
        expect(wrapper.find('figcaption').text().trim()).toBe('Author');
    });

    // Agregar <h3>{ counter }</h3> después del <hr /> para probar el useCounter
    test('Debe mostrar correctamente el counter', () => {
        useFetch.mockReturnValue({
            loading: false,
            error: null,
            data: [{
                quote_id: 1,
                quote: "Frase",
                author: "Author",
                series: "Serie"
            }]
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper.find('h3').text().trim()).toBe('10');
    });
    
})
