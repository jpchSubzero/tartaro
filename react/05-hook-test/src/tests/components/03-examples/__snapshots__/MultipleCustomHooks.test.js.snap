// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con MultipleCustomHooks Debe mostrar informaci√≥n 1`] = `
<Fragment>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <h3>
    10
  </h3>
  <figure
    className="text-end"
  >
    <blockquote
      className="blockquote"
    >
      <p
        className="mb-0"
      >
        Frase
      </p>
    </blockquote>
    <figcaption
      className="blockquote-footer"
    >
      Author
    </figcaption>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Siguiente frase
    </button>
  </figure>
</Fragment>
`;

exports[`Pruebas con MultipleCustomHooks Debe mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <h3>
    10
  </h3>
  <div
    className="alert alert-info text-center"
  >
    Cargando frase...
  </div>
</Fragment>
`;
